packages/story-editor/src/components/layout/index.js

```diff
@@ -81,24 +87,170 @@ function Layout({ header, children }) {
+    'web-stories'
+  );
+  if (allowedFileTypes.length) {
+    onSelectErrorMessage = sprintf(
+      /* translators: %s: list of allowed file types. */
+      __('Please choose only %s to insert into page.', 'web-stories'),
+      translateToExclusiveList(allowedFileTypes)
+    );
+  }
+
+  /**
+   * Insert element such image, video and audio into the editor.
+   *
+   * @param {Object} resource Resource object
+   * @param {string} thumbnailURL The thumbnail's url
+   * @return {null|*} Return onInsert or null.
+   */
+  const insertMediaElement = useCallback(
+    (resource, thumbnailURL) => {
+      resourceList.set(resource.id, {
+        url: thumbnailURL,
+        type: 'cached',
+      });
+      updateElementsById({
+        elementIds: [selectedElements?.[0]?.id],
+        properties: { resource },
+      });
+    },
+    [selectedElements, updateElementsById]
+  );
+
+  const handleMediaSelect = useCallback(
+    (resource) => {
+      try {
+        if (isTranscodingEnabled && canTranscodeResource(resource)) {
+          if (transcodableMimeTypes.includes(resource.mimeType)) {
+            optimizeVideo({ resource });
+          }
+
+          if (resource.mimeType === 'image/gif') {
+            optimizeGif({ resource });
+          }
+        }
+        // WordPress media picker event, sizes.medium.source_url is the smallest image
+        insertMediaElement(
+          resource,
+          resource.sizes?.medium?.source_url || resource.src
+        );
+
+        if (
+          !resource.local &&
+          allowedVideoMimeTypes.includes(resource.mimeType) &&
+          resource.isMuted === null
+        ) {
+          updateVideoIsMuted(resource.id, resource.src);
+        }
+      } catch (e) {
+        showSnackbar({
+          message: e.message,
+          dismissable: true,
+        });
+      }
+    },
+    [
+      allowedVideoMimeTypes,
+      insertMediaElement,
+      isTranscodingEnabled,
+      optimizeGif,
+      optimizeVideo,
+      showSnackbar,
+      transcodableMimeTypes,
+      updateVideoIsMuted,
+    ]
+  );
+
   return (
     <>
       <LayoutProvider>
         <ChecklistCheckpointProvider>
-          <HighlightsProvider>
-            <Editor zIndex={3}>
-              <CanvasProvider>
-                <RightClickMenuProvider>
-                  <Area area="lib">
-                    <Library />
-                  </Area>
-                  <Workspace header={header} />
-                  <RightClickMenu />
-                </RightClickMenuProvider>
-              </CanvasProvider>
-              {children}
-            </Editor>
-          </HighlightsProvider>
+          <MediaUploadProvider
+            title={__('Select media', 'web-stories')}
+            buttonInsertText={__('Insert media', 'web-stories')}
+            onSelect={handleMediaSelect}
+            onClose={resetWithFetch}
+            type={allowedMimeTypes}
+            onSelectErrorMessage={onSelectErrorMessage}
+          >
```

Comment:
I get worried about wrapping the whole editor in this. Would it be worth just wrapping sections of the editor where we need it?

maybe just wrapping up all the args and titles into something like this that we can use around the app?
```
<BaseMediaUploadProvider>
    // ...children
</BaseMediaUploadProvider>
```

Or does this sound silly? idk

Reply:
It just needs to wrap the quick action menu, so that should work

Reply:
The tricky thing is that it needs to live above the highlights provider because the quick actions live in the highlights provider

I agree though, it'd be nice to not wrap the whole app in this.

To fix this:
1. Refactor the quick actions into their own provider
2. Wrap just the quick action menu with the `QuickActionProvider` and the `MediaUploadProvider` 👍 

That would make this pr pretty massive though. I'm happy to follow up this PR with that refactor

Reply:
awesome. yea agreed. we can make it a code quality follow up